File: leetcode/container_with_most_water/container_with_most_water.cpp
/*
 * Time Complexity: O(N)
 *   - N is the length of the array.
 *
 * Space Complexity: O(1)
 */

#include <vector>
#include <algorithm>

using std::vector;
using std::max;
using std::min;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0, r = height.size()-1, ans = 0;

        while(l < r) {
            ans = max(ans, (r-l) * (min(height[l], height[r])));
            
            if(height[l] > height[r]) r--;
            else l++;
        }
        return ans;
    }
};

File: leetcode/container_with_most_water/container_with_most_water.java
package container_with_most_water;

class Solution {
    public int maxArea(int[] height) {
        int l = 0, r = height.length - 1, maxArea = 0;

        while (l < r) {
            maxArea = Math.max(maxArea, (r-l) * Math.min(height[l], height[r]));

            if (height[l] > height[r]) r--;
            else l++;
        }

        return maxArea;
    }
}
